bpf/lib/overloadable_xdp.h

```diff
@@ -167,12 +177,29 @@ ctx_set_encap_info(struct xdp_md *ctx __maybe_unused,
 		   __u32 node_id __maybe_unused,
 		   __u32 seclabel __maybe_unused,
 		   __u32 dstid __maybe_unused,
-		   __u32 vni __maybe_unused, int *ifindex __maybe_unused)
+		   __u32 vni __maybe_unused,
+		   void *opt __maybe_unused,
+		   __u32 opt_len __maybe_unused,
+		   bool is_ipv6 __maybe_unused,
+		   int *ifindex __maybe_unused)
 {
 	ctx_store_meta(ctx, CB_ENCAP_NODEID, bpf_ntohl(node_id));
 	ctx_store_meta(ctx, CB_ENCAP_SECLABEL, seclabel);
 	ctx_store_meta(ctx, CB_ENCAP_DSTID, dstid);
 	ctx_set_xfer(ctx, XFER_PKT_ENCAP);
+#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
+	if (opt) {
+		if (!is_ipv6) {
+			__be32 *addr = opt + sizeof(struct geneve_opt_hdr);
+			__be16 *port = opt + sizeof(struct geneve_opt_hdr) + sizeof(__be32);
+
+			ctx_store_meta(ctx, CB_ENCAP_PORT, *port);
+			ctx_store_meta(ctx, CB_ENCAP_ADDR, *addr);
+		} else {
+			return DROP_DSR_GENEVE_UNSUPP_IPV6_XDP;
```

Comment:
Did we consider at all attempting to let such packets through to the TC layer to forward the packets, albeit at a slower rate?

Reply:
(I don't consider this as a blocking comment right now, but it would be nice to at least consider this since it is introducing an incompatibility with some features that we otherwise support elsewhere)

Reply:
We support IPv4 in XDP using such a fallback mechanism, but we decided not to support IPv6 since the available data_meta size gets tricky when it comes to IPv6. Julian is working on the native XDP encap, which will lift the restriction for IPv6 and achieve optimal performance.